fvm/src/machine/limiter.rs

```diff
@@ -1,137 +1,126 @@
 // Copyright 2021-2023 Protocol Labs
 // SPDX-License-Identifier: Apache-2.0, MIT
-use wasmtime::ResourceLimiter;
 
 use crate::machine::NetworkConfig;
 
 /// Execution level memory tracking and adjustment.
-pub trait ExecMemory
-where
-    Self: Sized,
-{
+pub trait MemoryLimiter: Sized {
     /// Get a snapshot of the total memory required by the modules on the call stack so far.
-    fn curr_exec_memory_bytes(&self) -> usize;
+    fn memory_used(&self) -> usize;
+
+    /// Returns `true` if growing by `bytes` is allowed. Implement this memory to track and limit
+    /// memory usage.
+    fn grow_memory(&mut self, bytes: usize) -> bool;
 
     /// Push a new frame onto the call stack, and keep tallying up the current execution memory,
     /// then restore it to the current value when the frame is finished.
     fn with_stack_frame<T, G, F, R>(t: &mut T, g: G, f: F) -> R
     where
         G: Fn(&mut T) -> &mut Self,
         F: FnOnce(&mut T) -> R;
+
+    /// Grows an instance's memory from `from` to `to`. There's no need to manually implement this
+    /// unless you need to track instance metrics.
+    fn grow_instance_memory(&mut self, from: usize, to: usize) -> bool {
+        self.grow_memory(to.saturating_sub(from))
+    }
+
+    /// Grows an instance's table from `from` to `to` elements. There's no need to manually
+    /// implement this unless you need to track table metrics.
+    fn grow_instance_table(&mut self, from: u32, to: u32) -> bool {
+        // we charge 8 bytes per table element
+        self.grow_memory(to.saturating_sub(from).saturating_mul(8) as usize)
+    }
 }
 
 /// Limit resources throughout the whole message execution,
 /// across all Wasm instances.
-pub struct ExecResourceLimiter {
-    /// Maximum bytes that a single Wasm instance can use.
-    max_inst_memory_bytes: usize,
```

Comment:
The instance limit is enforced by wasmtime anyways. We _could_ enforce it here, but it's tricky when we take tables (and IPLD blocks) into account.

- For instance memory, wasmtime tells us how much memory is allocated when it requests more memory.
- For tables, wasmtime tells us how many elements are already in the table. But this doesn't help us apply an instance _memory_ limit.

So right now, this "instance" limit _only_ applies to the instance's _memory_.